Pester test runner: Default variables not applicable to release builds and needs more input validation

### VSTS Extension you are using
Pester test runner v8

### Where are you running it?
VSTS

### Version of Extension/Task
v8

### Expected behaviour and actual behaviour

When adding the task, it pre-populates some of the fields with defaults and some of them contain VSTS variables that are only available in build definitions, such as `$(Build.SourcesDirectory)`, which is not available in release definitions. Either a variables should be chosen that is available in both build and release or possibly a variable such as `$(System.DefaultWorkingDirectory)` should be chosen for release definitions.
![image](https://user-images.githubusercontent.com/9250262/42453260-186e1d4c-8384-11e8-9d02-ce46c6fb819d.png)
When it failed, it is was impossible to diagnose what went wrong, therefore more input validation would be useful

```
2018-07-09T13:15:23.1189871Z ##[section]Starting: Pester Test Runner
2018-07-09T13:15:23.1391553Z ==============================================================================
2018-07-09T13:15:23.1391819Z Task         : Pester Test Runner from Black Marble
2018-07-09T13:15:23.1392123Z Description  : Run Pester tests by either installing the latest version of Pester at run time (if possible) or using the version shipped with the task (4.3.1)
2018-07-09T13:15:23.1392382Z Version      : 8.1.2
2018-07-09T13:15:23.1392732Z Author       : Richard Fennell
2018-07-09T13:15:23.1393247Z Help         : Version: 8.1.2. [More Information](https://github.com/rfennell/vNextBuild/wiki/Pester-Task/)
2018-07-09T13:15:23.1393466Z ==============================================================================
2018-07-09T13:15:23.1708356Z Preparing task execution handler.
2018-07-09T13:15:23.7373209Z Executing the powershell script: D:\_Builds\RCSSV-EM20-005T.01\_tasks\Pester_31ef0033-64e3-4c55-b888-f446541474a6\8.1.2\Pester.ps1
2018-07-09T13:15:23.7490710Z ##[error]System.Management.Automation.ParameterBindingValidationException: Cannot validate argument on parameter 'resultsFile'. The "
2018-07-09T13:15:23.7499038Z             (Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')
2018-07-09T13:15:23.7499538Z         " validation script for the argument with value "$(Build.SourcesDirectory)\PesterPowerShellCmdletsSystemTests.XML" did not return a result of True. Determine why the validation script failed, and then try the command again. ---> System.Management.Automation.ValidationMetadataException: The "
2018-07-09T13:15:23.7499967Z             (Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')
2018-07-09T13:15:23.7500369Z         " validation script for the argument with value "$(Build.SourcesDirectory)\PesterPowerShellCmdletsSystemTests.XML" did not return a result of True. Determine why the validation script failed, and then try the command again.
2018-07-09T13:15:23.7501402Z    at System.Management.Automation.ValidateScriptAttribute.ValidateElement(Object element)
2018-07-09T13:15:23.7501757Z    at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
2018-07-09T13:15:23.7502217Z    --- End of inner exception stack trace ---
2018-07-09T13:15:23.7502399Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-07-09T13:15:23.7502644Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-07-09T13:15:23.7502887Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-07-09T13:15:23.7503174Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-07-09T13:15:23.7503468Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-07-09T13:15:23.7503760Z    at Microsoft.TeamFoundation.DistributedTask.Handlers.LegacyVSTSPowerShellHost.VSTSPowerShellHost.Main(String[] args)
2018-07-09T13:15:23.7511784Z ##[error]LegacyVSTSPowerShellHost.exe completed with return code: -1.
2018-07-09T13:15:23.7521745Z ##[section]Finishing: Pester Test Runner
```


### Steps to reproduce the problem

Thanks for this, I get it sorted this week. I think the System.WorkingDirectory exists in a build too so I'll swap to that once I check the docs.
________________________________
From: Christoph Bergmeister [MVP] <notifications@github.com>
Sent: Monday, July 9, 2018 2:29:42 PM
To: rfennell/vNextBuild
Cc: Subscribed
Subject: [rfennell/vNextBuild] Pester test runner: Default variables not applicable to release builds and needs more input validation (#350)

VSTS Extension you are using

Pester test runner v8

Where are you running it?

VSTS

Version of Extension/Task

v8

Expected behaviour and actual behaviour

When adding the task, it pre-populates some of the fields with defaults and some of them contain VSTS variables that are only available in build definitions, such as $(Build.SourcesDirectory), which is not available in release definitions. Either a variables should be chosen that is available in both build and release or possibly a variable such as $(System.DefaultWorkingDirectory) should be chosen for release definitions.
[image]<https://user-images.githubusercontent.com/9250262/42453260-186e1d4c-8384-11e8-9d02-ce46c6fb819d.png>
When it failed, it is was impossible to diagnose what went wrong, therefore more input validation would be useful

2018-07-09T13:15:23.1189871Z ##[section]Starting: Pester Test Runner
2018-07-09T13:15:23.1391553Z ==============================================================================
2018-07-09T13:15:23.1391819Z Task         : Pester Test Runner from Black Marble
2018-07-09T13:15:23.1392123Z Description  : Run Pester tests by either installing the latest version of Pester at run time (if possible) or using the version shipped with the task (4.3.1)
2018-07-09T13:15:23.1392382Z Version      : 8.1.2
2018-07-09T13:15:23.1392732Z Author       : Richard Fennell
2018-07-09T13:15:23.1393247Z Help         : Version: 8.1.2. [More Information](https://github.com/rfennell/vNextBuild/wiki/Pester-Task/)
2018-07-09T13:15:23.1393466Z ==============================================================================
2018-07-09T13:15:23.1708356Z Preparing task execution handler.
2018-07-09T13:15:23.7373209Z Executing the powershell script: D:\_Builds\RCSSV-EM20-005T.01\_tasks\Pester_31ef0033-64e3-4c55-b888-f446541474a6\8.1.2\Pester.ps1
2018-07-09T13:15:23.7490710Z ##[error]System.Management.Automation.ParameterBindingValidationException: Cannot validate argument on parameter 'resultsFile'. The "
2018-07-09T13:15:23.7499038Z             (Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')
2018-07-09T13:15:23.7499538Z         " validation script for the argument with value "$(Build.SourcesDirectory)\PesterPowerShellCmdletsSystemTests.XML" did not return a result of True. Determine why the validation script failed, and then try the command again. ---> System.Management.Automation.ValidationMetadataException: The "
2018-07-09T13:15:23.7499967Z             (Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')
2018-07-09T13:15:23.7500369Z         " validation script for the argument with value "$(Build.SourcesDirectory)\PesterPowerShellCmdletsSystemTests.XML" did not return a result of True. Determine why the validation script failed, and then try the command again.
2018-07-09T13:15:23.7501402Z    at System.Management.Automation.ValidateScriptAttribute.ValidateElement(Object element)
2018-07-09T13:15:23.7501757Z    at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
2018-07-09T13:15:23.7502217Z    --- End of inner exception stack trace ---
2018-07-09T13:15:23.7502399Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-07-09T13:15:23.7502644Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-07-09T13:15:23.7502887Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-07-09T13:15:23.7503174Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-07-09T13:15:23.7503468Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-07-09T13:15:23.7503760Z    at Microsoft.TeamFoundation.DistributedTask.Handlers.LegacyVSTSPowerShellHost.VSTSPowerShellHost.Main(String[] args)
2018-07-09T13:15:23.7511784Z ##[error]LegacyVSTSPowerShellHost.exe completed with return code: -1.
2018-07-09T13:15:23.7521745Z ##[section]Finishing: Pester Test Runner


Steps to reproduce the problem

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/rfennell/vNextBuild/issues/350>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQDqFgeUHeqmCruST0-Rox0dt7Hxh1CYks5uE1rGgaJpZM4VHtV->.

--- a/Extensions/Pester/readme.md
+++ b/Extensions/Pester/readme.md
@@ -44,3 +44,4 @@ Releases
     - Swap logging to use Write-Host to ensure it logs out by default. ([Fixes #320](https://github.com/rfennell/vNextBuild/issues/320))
     - Change Hashtable parsing function to use language parser to handle more cases. ([Fixes #321](https://github.com/rfennell/vNextBuild/issues/321))
 - 8.0.x - Removed complicated version loading logic and replaced with installing the latest version from the gallery if you're on PSv5+ or have PowerShellGet available. If neither of those are options then it will load the 4.3.1 version of Pester that ships with the task. ([PR#314](https://github.com/rfennell/vNextBuild/pull/314))
+- 8.1.x - Fixed the default working folder to System.DefaultWorkingDirectory as this is a variable available in both build and release. ([Fixes #350](https://github.com/rfennell/vNextBuild/issues/350))

--- a/Extensions/Pester/task/Pester.ps1
+++ b/Extensions/Pester/task/Pester.ps1
@@ -6,7 +6,12 @@ param
 
     [Parameter(Mandatory)]
     [ValidateScript( {
-            (Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')
+            if ((Test-Path (Split-Path $_ -Parent)) -and ($_.split('.')[-1] -eq 'xml')) {
+                $true
+            }
+            else {
+                Throw 'Path is invalid or results file does not end in .xml'
+            }
         })]
     [string]$resultsFile,
 

--- a/Extensions/Pester/task/Pester.ps1
+++ b/Extensions/Pester/task/Pester.ps1
@@ -10,7 +10,7 @@ param
                 $true
             }
             else {
-                Throw 'Path is invalid or results file does not end in .xml'
+                Throw "Path is invalid or results file does not end in .xml ($_)"
             }
         })]
     [string]$resultsFile,

--- a/Extensions/Pester/task/task.json
+++ b/Extensions/Pester/task/task.json
@@ -34,15 +34,15 @@
       "name": "scriptFolder",
       "type": "string",
       "label": "Scripts Folder or Script",
-      "defaultValue": "$(Build.SourcesDirectory)\\*",
+      "defaultValue": "$(System.DefaultWorkingDirectory)\\*",
       "required": true,
-      "helpMarkDown": "Folder to run scripts from e.g $(Build.SourcesDirectory)\\* or a script hashtable @{Path='$(Build.SourcesDirectory)'; Parameters=@{param1='111'; param2='222'}}"
+      "helpMarkDown": "Folder to run scripts from e.g $(System.DefaultWorkingDirectory)\\* or a script hashtable @{Path='$(System.DefaultWorkingDirectory)'; Parameters=@{param1='111'; param2='222'}}"
     },
     {
       "name": "resultsFile",
       "type": "string",
       "label": "Results File",
-      "defaultValue": "$(Build.SourcesDirectory)\\Test-Pester.XML",
+      "defaultValue": "$(System.DefaultWorkingDirectory)\\Test-Pester.XML",
       "required": true,
       "helpMarkDown": "Results File (nUnit format)"
     },

